### DISCUSSION TITLE
only device that performs 'accept' receives data

### DISCUSSION THREAD

**Discussion by TzuriTeshuba** [kenjdavidson/react-native-bluetooth-classic#186] [Q&A] [answered]
I downloaded the example project and encounter the issue where the accepting device successfully reads and write data, however the other side fails to receive data. the accepting side claims to have sent that data, but the onReceivedData(event) function is never called on the receiving side. is this a known issue? is there supposed to be an asymetry in the communication or any reason to expect this? Attached is your example code files if it helps. I had to change it a bit as many dependencies were no longer supported. the root component is 'AppCheck'. 

[src.zip](https://github.com/kenjdavidson/react-native-bluetooth-classic/files/8774684/src.zip)

**Comment by kenjdavidson** [comment, marked as answer]
Generally the issue is with the delimiter, the default is /n, if your device is different then you'll have to change it. If the configuration is correct then yes, the two way communication works just fine. It's all in the configuration for the inbound messages.

If you want to change code commit it back to your fork and I will compare. But at this time the app works for my purposes so I won't be pulling any changes.

At some point I'll need to update the dependencies, but I'll need to rewrite some of the library when that happens anyhow.

**Reply by TzuriTeshuba** [reply]
I will try out your suggestion and let the commmunity/future-issue-havers know if that solved it or something else. I mostly wanted to know if I should be expecting the issue or not. Thank you!

**Reply by kenjdavidson** [reply]
Right on.. keep me posted.  If you open an issue and provide:
- Your connection code
- Your reading code/listener code

and more importantly if possible:
- your device code
- your device API (if available)

it would help out a bunch for me to assist.  A lot of the issues for communication problems are:

- Needing byte array instead of delimited string
- Using the wrong encoding ASCII vs UTF
- Using the wrong delimiter `\r\n` instead of `\n`

things like that.

**Reply by gilgru23** [reply]
Hey tnx for helping, before we are looking for this properties I want to add to @TzuriTeshuba  words: This problem happens to us in multiple devices and not for a specific device. The device initiating the communication cannot receive messages (In the opposite direction everything works well). We tried with some devices to check the problem and always the one who initiates the communication can not get the messages even though these messages are sent(we see it in the console log). Please let us know if you still think the problem is related to delimiters or api of the device and we will bring you the details. Thank you very much :)

**Reply by kenjdavidson** [reply]
Oh. That seems unlikely then. So you're connecting two phones and the accepting one can always read but the connecting one never reads? That seems super weird. Once the acceptable is done the handing off is using the same code.

For example, my testing and usage incluces

1. Connecting Android to scanner. Both can read.
2. Connecting Android to tablet. Both can read.
3. Accepting tablet from Android. Both can read.

Whne I do device to device testing I use Serial Terminal app to limit any issues with this library. 

I suggest debuting the native code and see what's up with the message handling. It should be fairly straight forward. ,

**Reply by TzuriTeshuba** [reply]
just want to update, that we got the 2 way communication running properly, the solution was as simple as replacing the delimitter to '\n' from '\r' in both files (DeviceListScreen.js and ConnectionScreen.js) as you suggested. Thank you for the help!

**Reply by kenjdavidson** [reply]
Success!!! Glad to hear it.

**Comment by JMroue** [comment]
Evening good sir,

seems like i am having a similar problem.
i can send from the react-native app to the server via bluetooth socket (which is well received)
but i cannot read the received message from the server. 

i am connecting to a socket created using python in the code below:
`    
serverSocket = bluetooth.BluetoothSocket( bluetooth.RFCOMM )
    port = bluetooth.PORT_ANY
    serverSocket.bind(("", port))

    serverSocket.listen(1)
    print ("Listening for connections on port: ", port)

    port=serverSocket.getsockname()[1]
    bluetooth.advertise_service( serverSocket, "RaspberryPiService",
                        service_id = uuid,
                        service_classes = [ uuid, bluetooth.SERIAL_PORT_CLASS ],
                        profiles = [ bluetooth.SERIAL_PORT_PROFILE ] 
                        )
    
    inputSocket, address=serverSocket.accept()
    print ("Got connection with" , address)

    while True:
        data = inputSocket.recv(1024)
        parseReceived(data.decode())
        message = input("Enter response to send: ")
        if(message == 'b'):
            break
        sentBytes = inputSocket.send(message)
        print(f"sent bytes: {sentBytes}")


    inputSocket.close()
    serverSocket.close()  `
    
    then as suggested on the other side, i am connecting to the device using these options:
 `
const ConnectionOptions: StandardOptions = {
    secureSocket: false,
    delimiter: '\n'
} 
`
( i tried several \r as a delimiter and i have tried several charsets )
none of the read functions are working...

your help is much appreciated.

**Reply by kenjdavidson** [reply]
The original issue was solved by ensuring the delimiters matched.  If you're absolutely positive that they match and still having issues, you're going to have to do a little bit of debugging yourself.  I'm sorry, I don't have the time to setup your project and do the debugging myself at this time.

I'm not a python developer, but looking at your code (and knowing how other `input` functions work) there is no delimiter being sent by your server.

```
    // In all languages I've used, the command line input processing uses `\n` to end the input, 
    // and will not provide it as part of the content.  Is this how Python works?
    message = input("Enter response to send: ")
    if(message == 'b'):
        break
    sentBytes = inputSocket.send(message + "\n") // for example, can you do this?
```

ps. I get that // may not be a comment in Python (see I don't like python).

**Reply by JMroue** [reply]
Thanks a lot !!! This worked. 
So apparently in python, the input does not provide '\n' at the end.

seems like i do not like python also.
Thanks again.
